File: custom_addons/custom_widget/static/src/js/widget_expression.js
/** @odoo-module **/

import { Field } from "@web/views/fields/field";
import { Widget } from "@web/views/widgets/widget";

export class ExpressionWidget extends Widget {
    setup() {
        super.setup();
        this.isEditMode = this.props.mode === "edit";
        this.expression = this.props.value || "";
        this.result = this.evaluateExpression(this.expression);
    }

    evaluateExpression(expression) {
        try {
            // Evaluate the mathematical expression safely
            return eval(expression);
        } catch (error) {
            return "Invalid Expression";
        }
    }

    onInputChange(event) {
        const newValue = event.target.value;
        this.trigger("update", { value: newValue });
    }

    render() {
        if (this.isEditMode) {
            this.el.innerHTML = `<input type="text" value="${this.expression}" class="o_field_char" />`;
            this.el.querySelector("input").addEventListener("input", (e) => this.onInputChange(e));
        } else {
            this.el.innerHTML = `<span>${this.result}</span>`;
        }
    }
}

File: custom_addons/custom_widget/models/widget_expression.py
from odoo import models, fields

class MyModel(models.Model):
    _name = 'widget.model'
    _description = 'Widget Model'

    expression = fields.Char('Expression')
    result = fields.Float('Result', compute='_compute_result', store=True)

    def _compute_result(self):
        for record in self:
            try:
                # Evaluate the expression and store the result
                record.result = eval(record.expression)
            except Exception as e:
                record.result = 0  # If invalid expression, set result to 0
